File: app/src/main/java/com/instacart/android/challenges/MainActivityViewModel.java
package com.instacart.android.challenges;

import java.util.ArrayList;

import io.reactivex.annotations.Nullable;

public class MainActivityViewModel {

    interface UpdateListener {
        void onUpdate(ItemListViewState state);
    }

    private ItemListViewState itemListViewState;

    private UpdateListener listener;

    public MainActivityViewModel() {
        ArrayList<ItemRow> items = new ArrayList<>();
        items.add(new ItemRow("Cabbage"));
        items.add(new ItemRow("Apple"));
        items.add(new ItemRow("Bread"));

        itemListViewState = new ItemListViewState("Delivery Items", items);
    }

    public void setStateUpdateListener(@Nullable UpdateListener listener) {
        this.listener = listener;

        if (listener != null) {
            listener.onUpdate(itemListViewState);
        }
    }
}

File: app/src/main/java/com/instacart/android/challenges/MainActivityViewModel.kt
package com.instacart.android.challenges

import android.app.Application
import androidx.lifecycle.AndroidViewModel

class MainActivityViewModel(application: Application) : AndroidViewModel(application) {

    interface UpdateListener {
        fun onUpdate(state: ItemListViewState)
    }

    private var itemListViewState: ItemListViewState
    private var listener: UpdateListener? = null

    init {
        val items = listOf(
            ItemRow(name = "Cabbage"),
            ItemRow(name = "Apple"),
            ItemRow(name = "Bread")
        )

        itemListViewState = ItemListViewState("Delivery Items", items)
    }

    fun setStateUpdateListener(listener: UpdateListener?) {
        this.listener = listener

        listener?.onUpdate(itemListViewState)
    }
}
